llvm/lib/Object/SYCLBIN.cpp

```diff
@@ -0,0 +1,447 @@
+  GetMetadata(size_t ByteOffset, uint64_t MetadataSize) {
+    Expected<StringRef> BlobOrError = GetBinaryBlob(ByteOffset, MetadataSize);
+    if (!BlobOrError)
+      return BlobOrError.takeError();
+
+    std::string PropStr{BlobOrError->str()};
+    auto PropMemBuff = llvm::MemoryBuffer::getMemBuffer(PropStr);
+    auto ErrorOrProperties =
+        llvm::util::PropertySetRegistry::read(PropMemBuff.get());
+    if (!ErrorOrProperties)
+      return ErrorOrProperties.takeError();
+    return {std::move(*ErrorOrProperties)};
+  }
+};
+
+} // namespace
+
+SYCLBIN::SYCLBINDesc::SYCLBINDesc(BundleState State,
+                                  ArrayRef<SYCLBINModuleDesc> ModuleDescs) {
+  // TODO: Merge by properties, so overlap can live in the same abstract module.
+
+  // Write global metadata.
+  {
+    raw_svector_ostream GlobalMetadataOS(GlobalMetadata);
+    llvm::util::PropertySetRegistry GlobalMetadataProps;
+    GlobalMetadataProps.add(
+        llvm::util::PropertySetRegistry::SYCLBIN_GLOBAL_METADATA, "state",
+        static_cast<uint32_t>(State));
+    GlobalMetadataProps.write(GlobalMetadataOS);
+  }
+
+  // We currently create a single abstract module per split module.
+  // Some of these should be merged in the future.
+  size_t NumAMs = 0;
+  for (const SYCLBINModuleDesc &MD : ModuleDescs)
+    NumAMs += MD.SplitModules.size();
+  AbstractModuleDescs.reserve(NumAMs);
+
+  for (const SYCLBINModuleDesc &MD : ModuleDescs) {
+    for (const module_split::SplitModule &SM : MD.SplitModules) {
+      AbstractModuleDesc &AMD = AbstractModuleDescs.emplace_back();
+
+      // Write module metadata to the abstract module metadata.
+      raw_svector_ostream AMMetadataOS(AMD.Metadata);
+      SM.Properties.write(AMMetadataOS);
+
+      ImageDesc ID;
+      // Copy the filepath.
+      ID.FilePath = SM.ModuleFilePath;
+
+      // Create metadata and save the descriptor to the right collection.
+      raw_svector_ostream IDMetadataOS(ID.Metadata);
+      if (MD.ArchString.empty()) {
+        // If the arch string is empty, it must be an IR module.
+        llvm::util::PropertySetRegistry IRMMetadata;
+        // TODO: Determine type from the input.
+        IRMMetadata.add(
+            llvm::util::PropertySetRegistry::SYCLBIN_IR_MODULE_METADATA, "type",
+            /*SPIR-V*/ 0);
+        IRMMetadata.write(IDMetadataOS);
+        AMD.IRModuleDescs.emplace_back(std::move(ID));
+      } else {
+        // If the arch string is empty, it must be an native device code image.
+        llvm::util::PropertySetRegistry NDCIMetadata;
+        NDCIMetadata.add(llvm::util::PropertySetRegistry::
+                             SYCLBIN_NATIVE_DEVICE_CODE_IMAGE_METADATA,
+                         "arch", MD.ArchString);
+        NDCIMetadata.write(IDMetadataOS);
+        AMD.NativeDeviceCodeImageDescs.emplace_back(std::move(ID));
+      }
+    }
+  }
+}
+
+size_t SYCLBIN::SYCLBINDesc::getMetadataTableByteSize() const noexcept {
```

Comment:
nit.
LLVM is exception-less by design. Therefore, `noexcept` has no effect and brings only confusion.

Reply:
Ah, good point! Old habits for writing getters. I wonder if compilers still get some benefit from the promise of no exceptions, despite them never really being used in LLVM. Maybe they're smart enough to figure it out on their own.